### ISSUE TITLE
svmbir diverges when too many threads are used with small num_slices.

### ISSUE THREAD

**Issue by dyang37** [cabouman/svmbir#268]
Recently "bugzilla" is observed again on a problem with small number of slices, where divergence of the MBIR recon is observed with default number of threads on a problem with recon array size of 5x400x400.

I managed to reproduce this with 3D shepp Logan.
This can be reproduced by setting `num_slices=5`, `num_rows_cols=256`, and `max_resolutions=0` (**no multi-resolution**).
With the setting above, the maximum number of threads without breaking the code is 20. 
Here are the results:

**num_threads=20:**
convergence behavior:
![threads20](https://user-images.githubusercontent.com/7671736/200684653-15856d15-f2b6-4cca-9deb-617d5ac6d3c2.png)
final_recon:
![Screen Shot 2022-11-08 at 5 00 02 PM](https://user-images.githubusercontent.com/7671736/200684984-edfb0cd2-7621-4807-b57d-e71e9097e129.png)
(artifacts potentially due to small number of slices)

**num_threads=24:**
convergence behavior:
![Screen Shot 2022-11-08 at 5 02 37 PM](https://user-images.githubusercontent.com/7671736/200685376-c674745b-f574-4153-b315-3d1590f39111.png)
final recon:
![Screen Shot 2022-11-08 at 5 03 25 PM](https://user-images.githubusercontent.com/7671736/200685456-2f87d591-6912-4727-8fcb-cbbeb4dd1cac.png)

Again, I turned off multi-res to isolate the problem. However, the limit of the thread number could potentially change when multi-res is turned on.

**Comment by dyang37** [comment]
One more imgportant thing, the divergence behavior is only observed when **positivity=True**.
When positivity is turned off, MBIR converges as expected regardless the number of threads used.

**Comment by cabouman** [comment]
Two comments/questions:
- So it only diverges when `positivity=True`? That's crazy because positivity is supposed to help.
- How many cores did you have on the computer?

**Comment by dyang37** [comment]
> Two comments/questions:
> 
> * So it only diverges when `positivity=True`? That's crazy because positivity is supposed to help.
Yes.
> * How many cores did you have on the computer?
I used a server with 48 CPUs.

**Comment by gbuzzard** [comment]
Note that when positivity is false, then the c code will reduce the number of threads without warning.  So it's probably reducing to 20 or fewer threads in that case.

**Comment by dyang37** [comment]
Hi Xiao,
I still got the convergence issue when multi-res is disabled. The shepp-logan example above is produced with max_resolutions=0 and positivity=True.
Regards,
Diyu
________________________________

**Comment by XiaoWang-Github** [comment]
ok, then I have no clue then. 
This sounds really bizarre.
I have definitely run the super-voxel algorithm with much more than just 20 threads before. I didn't have any issue.

**Comment by cabouman** [comment]
I think we have a possible fix for the new Bugzilla. 
Replace lines 290-293 of recon3d.c with the following:
```
    max_of_num_slices_svdepth = ((num_slices < SVDEPTH) ? num_slices : SVDEPTH)
    i = (((Nx < Ny) ? Nx : Ny) / (2*SVLength+1)) * (max_of_num_slices_svdepth/SVDEPTH);
    max_threads = ( i < max_threads) ? i : max_threads ;
```

Previously, the problem is that 
```
SV_per_Z = ceil( num_slices/SVDEPTH )
```
So that we could allow overlap between SVs when the number of slices was not an integer multiple of the SV depth.

**Comment by dyang37** [comment]
PR to C-code is created here: https://github.com/HPImaging/sv-mbirct/pull/17
@sjkisner Could you please check that out when you have time?

**Comment by XiaoWang-Github** [comment]
Diyu, I went through the fix you proposed. 
It's great that you are investing the C code, but the fix you proposed doesn't make sense to me.

Let's look at the original C code that you want to change:
SV_per_Z = ceil( num_slices/SVDEPTH )

Suppose that we have 5 slices (num_slices=5), and the super-voxel depth is 4.
In that case, SV_per_Z which is the number of super-voxels along the z direction at a fixed X-Y plane coordinates, will be computed as 2 from the above equation ceil(5/4)=2.  And that's what it supposes to be. The first super-voxel will be across 4 slices, and the second super-voxel will only have 1 slice.
You don't want to use floor here, because that will skip the updates for the last slice.
I don't think the convergence issue has much to do with this line of code.

**Comment by sjkisner** [comment]
I pushed a fix in the **kisner/fix** branch.  I just limit the threads as in the case with no positivity constraint.  I may be able to relax this a bit by reducing the delay between pixel updates and error sinogram updates.

I'm not putting in the suggested change below because it results in max_threads=0 if the number of slices is less than SVDEPTH.
```
max_of_num_slices_svdepth = ((num_slices < SVDEPTH) ? num_slices : SVDEPTH)
i = (((Nx < Ny) ? Nx : Ny) / (2*SVLength+1)) * (max_of_num_slices_svdepth/SVDEPTH);
max_threads = ( i < max_threads) ? i : max_threads ;
```

**Comment by sjkisner** [comment]
>  final_recon: ![Screen Shot 2022-11-08 at 5 00 02 PM](https://user-images.githubusercontent.com/7671736/200684984-edfb0cd2-7621-4807-b57d-e71e9097e129.png) 
> (artifacts potentially due to small number of slices)

That inner halo is due to the inter-slice regularization.  The 3D phantom generated with a small number of slices makes it very coarse in Z.

I noticed we were requiring **b_interslice** to be positive. There isn't any reason it can't be set to zero, so I added a change to allow this.  If you set **b_interslice=0** in the test script that halo will goes away.
